src/types/profile-derived.js

```diff
@@ -187,8 +187,141 @@ export type SuffixOrderIndex = number;
+ *
+ * D -> B comes before A -> C, because B comes before C.
+ * D -> B comes after A -> B, because B == B and D comes after A.
+ * D -> B comes before A -> D -> B, because B == B, D == D, and "end of path" comes before A.
+ *
+ * ## Example
+ *
+ * ### Non-inverted call tree:
+ *
+ * ```
+ *   Tree            Left aligned    Right aligned        Reordered by suffix
+ * - [cn0] A      =  A            =            A [so0]    [so0] [cn0] A
+ *   - [cn1] B    =  A -> B       =       A -> B [so3]    [so1] [cn4] A <- A
+ *     - [cn2] A  =  A -> B -> A  =  A -> B -> A [so2] ↘↗ [so2] [cn2] A <- B <- A
+ *     - [cn3] C  =  A -> B -> C  =  A -> B -> C [so6] ↗↘ [so3] [cn1] B <- A
+ *   - [cn4] A    =  A -> A       =       A -> A [so1]    [so4] [cn5] B <- A <- A
+ *     - [cn5] B  =  A -> A -> B  =  A -> A -> B [so4]    [so5] [cn6] C <- A
+ *   - [cn6] C    =  A -> C       =       A -> C [so5]    [so6] [cn3] C <- B <- A
+ * ```
+ *
+ * ### Inverted call tree:
+ *
+ * ```
+ *                                                 Represents call paths ending in
+ * - [in0] A  (so:0..3)        =  A             =            ... A (cn0, cn4, cn2)
+ *   - [in1] A  (so:1..2)      =  A <- A        =       ... A -> A (cn4)
+ *   - [in2] B  (so:2..3)      =  A <- B        =       ... B -> A (cn2)
+ *     - [in3] A  (so:2..3)    =  A <- B <- A   =  ... A -> B -> A (cn2)
+ * - [in4] B  (so:3..5)        =  B             =            ... B (cn1, cn5)
+ *   - [in5] A  (so:3..5)      =  B <- A        =       ... A -> B (cn1, cn5)
+ *     - [in6] A  (so:4..5)    =  B <- A <- A   =  ... A -> A -> B (cn5)
+ * - [in7] C  (so:5..7)        =  C             =            ... C (cn6, cn3)
+ *   - [in8] A  (so:5..6)      =  C <- A        =       ... A -> C (cn6)
+ *   - [in9] B  (so:6..7)      =  C <- B        =       ... B -> C (cn3)
+ *     - [in10] A  (so:6..7)   =  C <- B <- A   =  ... A -> B -> C (cn3)
+ * ```
+ *
+ * In the suffix order, call paths become grouped in such a way that call paths
+ * which belong to the same *inverted* tree node (i.e. which share a suffix) end
+ * up ordered next to each other. This makes it so that a node in the inverted
+ * tree can refer to all its represented call paths with a single contiguous range.
+ *
+ * In this example, inverted tree node `in5` represents all call paths which end
+ * in A -> B. Both `cn1` and `cn5` do so; `cn1` is A -> B and `cn5` is A -> A -> B.
+ * In the suffix order, `cn1` and `cn5` end up next to each other, at positions
+ * `so3` and `so4`. This means that the two paths can be referred to via the suffix
+ * order index range 3..5.
+ *
+ * Suffix ordered call nodes: [0, 4, 2, 1, 5, 6, 3] (soX -> cnY)
+ * Suffix order indexes:      [0, 3, 2, 6, 1, 4, 5] (cnX -> soY)
+ *
+ * cnX:     Non-inverted call node index X
+ * soX:     Suffix order index X
+ * inX:     Inverted call node index X
+ * so:X..Y: Suffix order index range soX..soY (soY excluded)
  */
 export interface CallNodeInfoInverted extends CallNodeInfo {
```

Comment:
Fair. I can fold the interface and the implementation in a follow-up PR.

Reply:
Oh, I remember now: At the end of this PR, the two implementation classes are fully separate and don't inherit from each other. But the interfaces still inherit from each other. This way, both implementation classes can be treated as `CallNodeInfo` even though they're not related by subclassing.

But that doesn't justify the existence of the `CallNodeInverted` interface, so I could at least fold that into the inverted implementation class.

Reply:
Actually, if we keep the `CallNodeInfo` interface, and make `CallNodeInfoInverted` a concrete class, then I'm not sure how the `asInverted()` method on `CallNodeInfo` would work. What would the return type be? Would `profile-derived.js` be able to import the `CallNodeInfoInverted` type from `call-node-info.js`? I think that would cause a circular import dependency.

Reply:
I've added a commit which just moves everything into call-node-info.js and then combines the things there.